### PR TITLE
[Non-Atomic] Upgrade the codebase's utility system and correct fee determination and more....

### PR DESCRIPTION
**Pull request by plaxton**
**DO NOT MERGE! - STILL A WORK IN PROGRESS**

### What is the purpose of this pull request (PR)?

To upgrade the codebase's utility system and correct fee determination by abstraction along with certain 0.14 conformity changes

#### Was this PR tested and how?

Ubuntu 16.04 LTS on VMWare Workstation

### Why isn't this atomic

Many of the changes that were done **on point** required additional parent commits which required their parent commits which required certain codebase-wide changes which lead to this, I have tried to be as organized as possible by not doing superchange commits but there may be mistakes that I have attempted to fix

### What was exactly done in this commit?

- Introduction of -blockmintxfee, -incrementalrelayfee, -dustrelayfee params (fee abstraction)
- Implement **stub** bumpfee implementation
- Fix fee determination (as per BIP125) and above abstractions
- Update auxillary functions and design to match that of the above implementations
- Renaming main(.cpp/.h) to validation (.cpp/.h)
- Abstraction of boost::filesystem to std::filesystem
- Usage of std::function instead of boost::function
- Implement BCLog and upgrade LogPrint (not LogPrintf) assignments
- Change of 'dependancy' usage to latest miniupnpc due to CVE-2017-8798

### I hope this commit shall last the review
Like A Sir

### CODE REVIEW

File: src/activedynode.cpp [hunk 1]
```diff
@@ -20,9 +20,9 @@ CActiveDynode activeDynode;
 
 void CActiveDynode::ManageState()
 {
-    LogPrint("Dynode", "CActiveDynode::ManageState -- Start\n");
+    LogPrint(BCLog::DYNODE, "CActiveDynode::ManageState -- Start\n");
```
Comment (spencerlievens) [inline src/activedynode.cpp]: Change BCLog to DYNLog throughout.

File: src/dynode-payments.cpp [hunk 2]
```diff
@@ -51,18 +51,18 @@ bool IsBlockValueValid(const CBlock& block, int nBlockHeight, CAmount blockRewar
         if(nBlockHeight >= consensusParams.nBudgetPaymentsStartBlock &&
             nOffset < consensusParams.nBudgetPaymentsWindowBlocks) {
             if(dynodeSync.IsSynced() && !sporkManager.IsSporkActive(SPORK_13_OLD_SUPERBLOCK_FLAG)) {
-                LogPrint("gobject", "IsBlockValueValid -- Client synced but budget spork is disabled, checking block value against block reward\n");
+                LogPrint(BCLog::GOBJECT, "IsBlockValueValid -- Client synced but budget spork is disabled, checking block value against block reward\n");
```
Comment (spencerlievens) [inline src/dynode-payments.cpp]: Change to DYNLog::

File: src/dynode-sync.cpp [hunk 3]
```diff
@@ -359,7 +359,7 @@ void CDynodeSync::ProcessTick()
             // MNLIST : SYNC DYNODE LIST FROM OTHER CONNECTED CLIENTS
 
             if(nRequestedDynodeAssets == DYNODE_SYNC_LIST) {
-                LogPrint("Dynode", "CDynodeSync::ProcessTick -- nTick %d nRequestedDynodeAssets %d nTimeLastDynodeList %lld GetTime() %lld diff %lld\n", nTick, nRequestedDynodeAssets, nTimeLastDynodeList, GetTime(), GetTime() - nTimeLastDynodeList);
+                LogPrint(BCLog::DYNODE, "CDynodeSync::ProcessTick -- nTick %d nRequestedDynodeAssets %d nTimeLastDynodeList %lld GetTime() %lld diff %lld\n", nTick, nRequestedDynodeAssets, nTimeLastDynodeList, GetTime(), GetTime() - nTimeLastDynodeList);
```
Comment (spencerlievens) [inline src/dynode-sync.cpp]: Change to DYNLog::

File: src/dynode.cpp [hunk 4]
```diff
@@ -171,7 +171,7 @@ void CDynode::Check(bool fForce)
     if(!fForce && (GetTime() - nTimeLastChecked < DYNODE_CHECK_SECONDS)) return;
     nTimeLastChecked = GetTime();
 
-   LogPrint("Dynode", "CDynode::Check -- Dynode %s is in %s state\n", vin.prevout.ToStringShort(), GetStateString());
+   LogPrint(BCLog::DYNODE, "CDynode::Check -- Dynode %s is in %s state\n", vin.prevout.ToStringShort(), GetStateString());
```
Comment (spencerlievens) [inline src/dynode.cpp]: Change to DYNLog::

File: src/dynodeman.cpp [hunk 5]
```diff
@@ -128,7 +128,7 @@ bool CDynodeMan::Add(CDynode &dn)
 
     CDynode *pdn = Find(dn.vin);
     if (pdn == NULL) {
-        LogPrint("Dynode", "CDynodeMan::Add -- Adding new Dynode: addr=%s, %i now\n", dn.addr.ToString(), size() + 1);
+        LogPrint(BCLog::DYNODE, "CDynodeMan::Add -- Adding new Dynode: addr=%s, %i now\n", dn.addr.ToString(), size() + 1);
```
Comment (spencerlievens) [inline src/dynodeman.cpp]: Change to DYNLog::

File: src/fs.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_FS_H
```
Comment (spencerlievens) [inline src/fs.h]: Change to DYNAMIC_FS_H

File: src/fs.h [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+/** Filesystem operations and types */
+namespace fs = boost::filesystem;
+
+/** Bridge operations to C stdio */
+namespace fsbridge {
+    FILE *fopen(const fs::path& p, const char *mode);
+    FILE *freopen(const fs::path& p, const char *mode, FILE *stream);
+};
+
+#endif
```
Comment (spencerlievens) [inline src/fs.h]: Add // DYNAMIC_FS_H

File: src/governance-classes.cpp [hunk 8]
```diff
@@ -158,7 +158,7 @@ bool CGovernanceTriggerManager::AddNewTrigger(uint256 nHash)
 
 void CGovernanceTriggerManager::CleanAndRemove()
 {
-    LogPrint("gobject", "CGovernanceTriggerManager::CleanAndRemove -- Start\n");
+    LogPrint(BCLog::GOBJECT, "CGovernanceTriggerManager::CleanAndRemove -- Start\n");
```
Comment (spencerlievens) [inline src/governance-classes.cpp]: Change to DYNLog::

File: src/governance-object.cpp [hunk 9]
```diff
@@ -111,7 +111,7 @@ bool CGovernanceObject::ProcessVote(CNode* pfrom,
             LogPrintf(ostr.str().c_str());
         }
         else {
-            LogPrint("gobject", ostr.str().c_str());
+            LogPrint(BCLog::GOBJECT, ostr.str().c_str());
```
Comment (spencerlievens) [inline src/governance-object.cpp]: Change to DYNLog::

File: src/governance-object.h [hunk 10]
```diff
@@ -318,12 +318,12 @@ class CGovernanceObject
         READWRITE(vchSig);
         if(nType & SER_DISK) {
             // Only include these for the disk file format
-            LogPrint("gobject", "CGovernanceObject::SerializationOp Reading/writing votes from/to disk\n");
+            LogPrint(BCLog::GOBJECT, "CGovernanceObject::SerializationOp Reading/writing votes from/to disk\n");
```
Comment (spencerlievens) [inline src/governance-object.h]: Change to DYNLog::

File: src/governance-vote.cpp [hunk 11]
```diff
@@ -257,27 +257,27 @@ bool CGovernanceVote::Sign(CKey& keyDynode, CPubKey& pubKeyDynode)
 bool CGovernanceVote::IsValid(bool fSignatureCheck) const
 {
     if(nTime > GetTime() + (60*60)) {
-        LogPrint("gobject", "CGovernanceVote::IsValid -- vote is too far ahead of current time - %s - nTime %lli - Max Time %lli\n", GetHash().ToString(), nTime, GetTime() + (60*60));
+        LogPrint(BCLog::GOBJECT, "CGovernanceVote::IsValid -- vote is too far ahead of current time - %s - nTime %lli - Max Time %lli\n", GetHash().ToString(), nTime, GetTime() + (60*60));
```
Comment (spencerlievens) [inline src/governance-vote.cpp]: Change to DYNLog::

File: src/governance.cpp [hunk 12]
```diff
@@ -132,7 +132,7 @@ void CGovernanceManager::ProcessMessage(CNode* pfrom, std::string& strCommand, C
             if(nProp == uint256()) {
                 if(netfulfilledman.HasFulfilledRequest(pfrom->addr, NetMsgType::DNGOVERNANCESYNC)) {
                     // Asking for the whole list multiple times in a short period of time is no good
-                    LogPrint("gobject", "DNGOVERNANCESYNC -- peer already asked me for the list\n");
+                    LogPrint(BCLog::GOBJECT, "DNGOVERNANCESYNC -- peer already asked me for the list\n");
```
Comment (spencerlievens) [inline src/governance.cpp]: Change to DYNLog::

File: src/governance.h [hunk 13]
```diff
@@ -300,7 +300,7 @@ class CGovernanceManager
     {
         LOCK(cs);
 
-        LogPrint("gobject", "Governance object manager was cleared\n");
+        LogPrint(BCLog::GOBJECT, "Governance object manager was cleared\n");
```
Comment (spencerlievens) [inline src/governance.h]: Change to DYNLog::

File: src/httprpc.cpp [hunk 14]
```diff
@@ -232,7 +232,7 @@ static bool InitRPCAuthentication()
 
 bool StartHTTPRPC()
 {
-    LogPrint("rpc", "Starting HTTP RPC server\n");
+    LogPrint(BCLog::RPC, "Starting HTTP RPC server\n");
```
Comment (spencerlievens) [inline src/httprpc.cpp]: Change to DYNLog::

File: src/httpserver.cpp [hunk 15]
```diff
@@ -212,7 +212,7 @@ static bool InitHTTPAllowList()
     std::string strAllowed;
     for (const CSubNet& subnet : rpc_allow_subnets)
         strAllowed += subnet.ToString() + " ";
-    LogPrint("http", "Allowing HTTP connections from: %s\n", strAllowed);
+    LogPrint(BCLog::HTTP, "Allowing HTTP connections from: %s\n", strAllowed);
```
Comment (spencerlievens) [inline src/httpserver.cpp]: Change to DYNLog::

File: src/init.cpp [hunk 16]
```diff
@@ -348,7 +348,7 @@ bool static Bind(const CService &addr, unsigned int flags) {
 void OnRPCStopped()
 {
     cvBlockChange.notify_all();
-    LogPrint("rpc", "RPC stopped.\n");
+    LogPrint(BCLog::RPC, "RPC stopped.\n");
```
Comment (spencerlievens) [inline src/init.cpp]: Change to DYNLog::

File: src/instantsend.cpp [hunk 17]
```diff
@@ -97,7 +97,7 @@ bool CInstantSend::ProcessTxLockRequest(const CTxLockRequest& txLockRequest)
         if(it != mapVotedOutpoints.end()) {
             BOOST_FOREACH(const uint256& hash, it->second) {
                 if(hash != txLockRequest.GetHash()) {
-                    LogPrint("instantsend", "CInstantSend::ProcessTxLockRequest -- Double spend attempt! %s\n", txin.prevout.ToStringShort());
+                    LogPrint(BCLog::INSTANTSEND, "CInstantSend::ProcessTxLockRequest -- Double spend attempt! %s\n", txin.prevout.ToStringShort());
```
Comment (spencerlievens) [inline src/instantsend.cpp]: Change to DYNLog::

**Comment by spencerlievens** [comment]
Please look at review comments. Change all BCLog instances to DYNLog. Apart from that, it looks good.

utACK.

**Comment by plaxton** [comment]
Change in favour of shifting to Version 1.5.0.0

**Comment by crackcomm** [comment]
Tracking wallet address needs check because `getbalance` reports zero.

**Pull request closed without merging**